qiskit/transpiler/synthesis/aqc/cnot_structures.py

```diff
@@ -0,0 +1,325 @@
+
+        cnots = np.zeros((2, depth), dtype=np.int64)
+        for i in range(depth):
+            cnots[0, i] = (i + 0) % num_qubits
+            cnots[1, i] = (i + 1) % num_qubits
+        cnots += 1  # 1-based index
+        return cnots
+
+    else:
+        raise ValueError(
+            f"unknown type of CNOT-network layout, expects one of {get_network_layouts()}, "
+            f"got {network_layout}"
+        )
+
+
+def generate_random_cnots(num_qubits: int, depth: int, set_depth_limit: bool = True) -> np.ndarray:
+    """
+    Generates a random CNot network for debugging and testing. Note: there's 1-based index.
+
+    Args:
+        num_qubits: number of qubits.
+        depth: depth of the network; it will be bounded by the lower limit (function
+            ``lower_limit()``), if exceeded.
+        set_depth_limit: apply theoretical upper limit on cnot circuit depth.
+
+    Returns:
+        A matrix of size ``(2, N)``, where each column contains two distinct indices from
+            the range ``[1 ... num_qubits]``;
+    """
+    assert isinstance(num_qubits, (np.int64, int)) and 1 <= num_qubits <= 16
+    # pylint: disable=misplaced-comparison-constant
+    assert isinstance(depth, (np.int64, int)) and 1 <= depth
+    assert isinstance(set_depth_limit, bool)
+
+    if set_depth_limit:
+        depth = min(depth, lower_limit(num_qubits))
+    cnots = np.tile(np.arange(num_qubits).reshape(num_qubits, 1), depth)
+    for i in range(depth):
+        np.random.shuffle(cnots[:, i])
+    cnots = cnots[0:2, :].copy()
+    cnots += 1  # TODO: 1-based index
+    return cnots
+
+
+def get_connectivity(num_qubits: int, connectivity: str) -> dict:
+    """
+    Generates connectivity structure between qubits.
+
+    Args:
+        num_qubits: number of qubits.
+        connectivity: type of connectivity structure, ``{"full", "line", "star"}``.
+
+    Returns:
+        dictionary of allowed links between qubits.
+
+    Raises:
+         ValueError: if unsupported type of CNOT-network layout is passed.
+    """
+    assert isinstance(num_qubits, (np.int64, int)) and num_qubits > 0
+    assert isinstance(connectivity, str)
+
+    if num_qubits == 1:
+        links = {1: [1]}
+
+    elif connectivity == "full":
+        # Full connectivity between qubits.
+        links = {i + 1: [j + 1 for j in range(num_qubits)] for i in range(num_qubits)}
+
+    elif connectivity == "line":
+        # Every qubit is connected to its immediate neighbours only.
+        links = {i + 1: [i, i + 1, i + 2] for i in range(1, num_qubits - 1)}
+        links[1] = [1, 2]
+        links[num_qubits] = [num_qubits - 1, num_qubits]
+
+    elif connectivity == "star":
+        # Every qubit is connected to the first one only.
+        links = {i + 1: [1, i + 1] for i in range(1, num_qubits)}
+        links[1] = [j + 1 for j in range(num_qubits)]
+
+    else:
+        raise ValueError(
```

Comment:
maybe a custom exception here?

Reply:
A good question. What are the rules for exception usage, e.g. when re-use an existing one, when create a sub-class, etc?